enacts/calc.py

```diff
@@ -25,13 +25,37 @@ def read_zarr_data(zarr_path):
 
 # Growing season functions
 
+def water_balance_step(
+    sm_history,
+    peffective,
+    et,
+    taw,
+    reduce=True,
+    time_dim="T",
+):
+    sm = (
+        # if I don't expand dims then time_dim no more a dim, just a coordinate
+        # and following up assign_coords goes bananas -- weird
+        sm_history.isel({time_dim: -1}).expand_dims(dim=time_dim) + peffective - et
+    ).clip(min=0, max=taw).assign_coords(
+        {time_dim: [(sm_history[time_dim][-1] + np.timedelta64(1, "D")).values]}
+    )
+    if not reduce:
+        #at this point, this assumes that
+        #peffective, et and taw don't add new dimensions
+        #through broadcasting to sm_history
+        #so that the concat can happen
+        sm = xr.concat([sm_history, sm], time_dim)
+    return sm
```

Comment:
I would prefer to have the following simple function that just describes the calculation itself
```
def water_balance_step(sm_prev, peffective, et, taw):
    return (sm_prev + peffective - et).clip(min=0, max=taw)
```
and put the burden on the caller to manage the bookkeeping about what to do with the result (store it in a big array vs. just pass it to the next iteration).

I'm not sure that preference will be compatible with an eventual reimplementation in C/Cython, though. So if you don't agree with my aesthetic opinion, feel free to ignore it, at least for now.

Unrelated: in the case where we need to keep the entire history, pre-allocating the whole array and writing slices into it will probably be significantly faster than incrementally growing the array by calling `concat` on each iteration.

Reply:
So rather something like what I've just pushed?

Reply:
I'll do the pre-allocated array after we're good on where to put the burden of reduction decision.

Reply:
Yes, for my first comment. The comment about preallocating the array still stands.

Reply:
Before I do that, a comment that preallocating the array means I will have to manually broadcast everything to have the right sm array from the start (which resulted in most of the lines of code in the "good" water balance function).

whereas here, even though [xr.concat](https://docs.xarray.dev/en/stable/generated/xarray.concat.html) claims that

"Each object is expected to consist of variables and coordinates with matching shapes except for along the concatenated dimension."

sm_i is concatenated to soil_moisture in the first iteration in test_water_balance2() even though sminit has no X dimension but precip does... so apparently broadcasting things smoothly.

Reply:
Or maybe I run the first iteration of water_balance_step so that it does the broadcasting, then I allocate with the proper T dim, then I loop from 1 to T.size?

Reply:
Whatever seems easiest. If it's going to make the code harder to read, we should verify that it actually helps performance significantly. But you can't verify that without making the change first.

Reply:
how about that.

I don't know why but all the initialization and broadcasting problems are fading away...